tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_gemv_kk.hpp

```diff
@@ -0,0 +1,228 @@
+  }
+  if ( A.extent(0) != z.extent(0) ){
+    throw std::runtime_error("KokkosBlas: matrix_vector_product: A.extent(0) != z.extent(0) ");
+  }
+
+  // this print is detected in the tests
+#if defined KOKKOS_STDBLAS_ENABLE_TESTS
+  std::cout << "updating_matrix_vector_product: kokkos impl\n";
+#endif
+
+  auto A_view = Impl::mdspan_to_view(A);
+  auto x_view = Impl::mdspan_to_view(x);
+  auto y_view = Impl::mdspan_to_view(y);
+  auto z_view = Impl::mdspan_to_view(z);
+
+  // FRIZZI: we need to improve this or maybe we should
+  // just do two ops: fist y->z, and then gemv directly
+  auto ex = ExeSpace();
+  Kokkos::parallel_for(Kokkos::RangePolicy(ex, 0, z_view.extent(0)),
+		       KOKKOS_LAMBDA (const std::size_t & i)
+		       {
+			 typename decltype(z_view)::value_type z_i = {};
+			 for (std::size_t j=0; j<A_view.extent(1); ++j){
+			   z_i += y_view(i) + A_view(i,j) * x_view(j);
+			 }
+			 z_view(i) = z_i;
+		       });
+
+  //fence message when using latest kokkos:
+  ex.fence();
+  // ex.fence("KokkosStdBlas::gemv: fence after operation");
+}
+
+//
+// the following specialization is here to avoid breaking
+// the example code started by CTrott in this file:
+//    stdBLAS/examples/kokkos-based/matrix_vector_product_kokkos.cpp
+// but we need to change this later after we properly handle accessors
+//
+template<//class ExecSpace,
+         class ElementType_A,
+         std::experimental::extents<>::size_type numRows_A,
+         std::experimental::extents<>::size_type numCols_A,
+         class Layout_A,
+         class Accessor_A,
+         class ElementType_x,
+         std::experimental::extents<>::size_type ext_x,
+         class Layout_x,
+         class Accessor_x,
+         class ElementType_y,
+         std::experimental::extents<>::size_type ext_y,
+         class Layout_y,
+         class Accessor_y>
+std::enable_if_t<
+  !std::is_same<Accessor_A, std::experimental::default_accessor<ElementType_A>>::value
+  and !std::is_same<Accessor_x, std::experimental::default_accessor<ElementType_x>>::value
+  and !std::is_same<Accessor_y, std::experimental::default_accessor<ElementType_y>>::value
+  >
+matrix_vector_product(kokkos_exec<>,
+		      std::experimental::mdspan<
+		        ElementType_A,
+		        std::experimental::extents<numRows_A, numCols_A>,
+		        Layout_A, Accessor_A
+		      > A,
+		      std::experimental::mdspan<
+		        ElementType_x,
+		        std::experimental::extents<ext_x>,
+		        Layout_x,
+		        Accessor_x
+		      > x,
+		      std::experimental::mdspan<
+		        ElementType_y,
+		        std::experimental::extents<ext_y>,
+		        Layout_y,
+		        Accessor_y
+		      > y)
+{
+  auto alpha = gemvimpl::get_scaling_factor(A.accessor())  *
+               gemvimpl::get_scaling_factor(x.accessor());
+  KokkosBlas::gemv("N", alpha, Impl::mdspan_to_view(A),
+		   Impl::mdspan_to_view(x), 0.0,
```

Comment:
Would you consider making sure that beta has the same type as `alpha` (or some other appropriate type), instead of just using `double` (i.e., `0.0`) here?

Reply:
yes, sorry, this slipped! thanks for bringing this up